src/mlpack/methods/linear_svm/linear_svm_function_impl.hpp

```diff
@@ -0,0 +1,511 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_LINEAR_SVM_LINEAR_SVM_FUNCTION_IMPL_HPP
+#define MLPACK_METHODS_LINEAR_SVM_LINEAR_SVM_FUNCTION_IMPL_HPP
+
+#include <mlpack/core/math/make_alias.hpp>
+#include <mlpack/core/math/shuffle_data.hpp>
+
+// In case it hasn't been included yet.
+#include "linear_svm_function.hpp"
+
+namespace mlpack {
+namespace svm {
+
+template <typename MatType>
+LinearSVMFunction<MatType>::LinearSVMFunction(
+    const MatType& dataset,
+    const arma::Row<size_t>& labels,
+    const size_t numClasses,
+    const double lambda,
+    const double delta,
+    const bool fitIntercept) :
+    dataset(math::MakeAlias(const_cast<MatType&>(dataset), false)),
+    numClasses(numClasses),
+    lambda(lambda),
+    delta(delta),
+    fitIntercept(fitIntercept)
+{
+  InitializeWeights(initialPoint, dataset.n_rows, numClasses, fitIntercept);
+  initialPoint *= 0.005;
+
+  // Calculate the label matrix.
+  GetGroundTruthMatrix(labels, groundTruth);
+}
+
+/**
+ * Initializes parameter weights to random values taken from a scaled standard
+ * normal distribution. The weights cannot be initialized to zero, as that will
+ * lead to each class output being the same.
+ */
+template <typename MatType>
+void LinearSVMFunction<MatType>::InitializeWeights(
+    arma::mat &weights,
+    const size_t featureSize,
+    const size_t numClasses,
+    const bool fitIntercept)
+{
+  // Initialize values to 0.005 * r. 'r' is a matrix of random values taken from
+  // a Gaussian distribution with mean zero and variance one.
+  if (fitIntercept)
+    weights.randn(featureSize + 1, numClasses);
+  else
+    weights.randn(featureSize, numClasses);
+  weights *= 0.005;
+}
+
+/**
+ * This is equivalent to applying the indicator function to the training
+ * labels. The output is in the form of a matrix, which leads to simpler
+ * calculations in the Evaluate() and Gradient() methods.
+ */
+template <typename MatType>
+void LinearSVMFunction<MatType>::GetGroundTruthMatrix(
+    const arma::Row<size_t>& labels,
+    arma::sp_mat& groundTruth)
+{
+  // Calculate the ground truth matrix according to the labels passed. The
+  // ground truth matrix is a matrix of dimensions 'numClasses * numExamples',
+  // where each column contains a single entry of '1', marking the label
+  // corresponding to that example.
+
+  // Row pointers and column pointers corresponding to the entries.
+  arma::uvec rowPointers(labels.n_elem);
+  arma::uvec colPointers(labels.n_elem + 1);
+
+  // Row pointers are the labels of the examples, and column pointers are the
+  // number of cumulative entries made uptil that column.
+  for (size_t i = 0; i < labels.n_elem; i++)
+  {
+    rowPointers(i) = labels(i);
+    colPointers(i + 1) = i + 1;
+  }
```

Comment:
Here's the problem.  `colPointers[0]` is never set to anything.  It should actually be set to 0, so you can add `colPointers[0] = 0` before the loop.  And now here is the long story of how I got here.

I could see that the most recent commit could definitely cause `LinearSVMTest` to fail.  I thought a good first step towards solving this would be simply to print what `groundTruth.cols(firstId, lastId)` is in `Evaluate()`.  So I did that, and strangely found that `groundTruth.cols(firstId, lastId)` (with `firstId == lastId == 0`) said it was empty even though if I printed the whole `groundTruth` I saw

```
[matrix size: 2x3; n_nonzero: 3; density: 50.00%]

     (1, 0)         1.0000
     (0, 1)         1.0000
     (1, 2)         1.0000
```

Immediately I thought that perhaps there is a bug in Armadillo.  Specifically, this problem would be most likely in the sparse subview iterators, since the matrix printing functionality should simply be looping over the subview with a `const_iterator` or something.  So I opened up `armadillo_bits/SpSubview_iterators_meat.hpp`, which is one of my least favorite files in the world because it turns out that iterating over a subview of a sparse matrix stored in CSC format is kind of a disaster of complex code.  I added `cout` calls all over the place just so I could get an idea of what was going on, but to my surprise, there was actually no output when I ran the code!

So at this point I pulled out `gdb` and stepped into the call of `groundTruth.cols(firstId, lastId)`.  This call to `cols()` creates an `SpSubview`, so I stepped through the first constructor in `armadillo_bits/SpSubview_meat.hpp`.  That file has the following code:

```
  // There must be a O(1) way to do this
  uword lend     = m.col_ptrs[in_col1 + in_n_cols];
  uword lend_row = in_row1 + in_n_rows;
  uword count   = 0;

  for(uword i = m.col_ptrs[in_col1]; i < lend; ++i)
    {
    const uword m_row_indices_i = m.row_indices[i];

    const bool condition = (m_row_indices_i >= in_row1) && (m_row_indices_i < lend_row);

    count += condition ? uword(1) : uword(0);
    }

  access::rw(n_nonzero) = count;
```

So, when I stepped through this with gdb, I found that `count` ended up being 0 for our single-column submatrix.  But we can clearly see from the full `groundTruth` matrix that it should be 1---the code just counts the number of elements in the subview.  So I looked at the loop initialization variable `m.col_ptrs[in_col1]`---where `in_col1` was 0, and I found that actually `m.col_ptrs[0]` was some gigantic uninitialized value!  But the rest of `m.col_ptrs` were fine.  (The column pointers are part of the CSC format, and `m` here is the full sparse matrix, or `groundTruth` in our case.)

I took a look through the different `SpMat` constructors, to see if I could pinpoint any clear issue or something to look further into---but I saw they all correctly set every `col_ptrs` element to 0 whenever the matrix was constructed or resized.

So next I thought that a possibility was that `m.col_ptrs[0]` could give uninitialized memory if `col_ptrs` was never set to a size (i.e. if the matrix had 0 columns).  This made me look at where `groundTruth` was created, but I saw in the calls to `arma::sp_mat()` (like below) that the size was set reasonably.

Then I realized that we were setting the row indices and column pointers manually in this function, and suddenly it became clear as soon as I looked at this loop what was going on. :)

Once you add the line I suggested to the top of the loop, now the tests pass. :+1:

Reply:
Thanks for digging into this. I added the changes that you suggested and added them, and they worked perfectly. 
The fact that printing the entire `groundTruth` was not giving any issue but using `cols` over it was giving weird issues maybe tricked us into thinking that, maybe there was something wrong with Armadillo, but the actual issue was with the function where we manually initialised `groundTruth`.
I'll read your comment a few more times, to understand it properly, since you did alot of stuff here (And the way you debugged was cool :smile_cat:).

**P.S.**: Also, I made the same changes to the same function inside `SoftmaxRegression`, since I took the function from there (and it is not causing issue right now), and it could prevent similar issue from happening from the future if any changes are made to `SoftmaxRegression` or if anyone uses the same function in their implementation.

Reply:
Wonderful, thanks for making the changes there also. :+1: